qwerty_ast/src/typecheck.rs

```diff
@@ -160,17 +216,37 @@ pub fn typecheck_expr(expr: &Expr, env: &mut TypeEnv) -> Result<Type, TypeError>
 
         Expr::UnitLiteral { dbg: _ } => Ok(Type::UnitType),
 
-        Expr::Adjoint { func, dbg: _ } => {
+        Expr::Adjoint { func, dbg } => {
             // Adjoint should be a function type (unitary/quantum), not classical.
             let func_ty = typecheck_expr(func, env)?;
-            // TODO: Enforce Qwerty adjoint typing rules.
-            Ok(func_ty)
+
+            match func_ty {
+                Type::RevFuncType { in_out_ty } => Ok(Type::RevFuncType { in_out_ty }),
```

Comment:
I didn't communicate this to you properly, but I am thinking that the operand `e` of `~e` (in terms of your data structure, the `func` of an `Adjoint`) should have an `in_ty` of `qubit[m]`, where $m > 0$. Like this rule:

![Image](https://github.com/user-attachments/assets/a6e803cb-4d9e-4804-8250-edaf94fcd357)

Some of the MLIR work downstream expects this because it made the implementation dramatically easier, so it would be nice if the type system in the frontend could enforce that.

Reply:
Looks like I didn't reply to this either, sorry for that.